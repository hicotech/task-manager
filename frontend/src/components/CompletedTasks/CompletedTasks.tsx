import { TaskList } from '@task-manager/shared-library';
import styled from 'styled-components';
import Circle from '../Icon/Circle';
import Polygon from '../Icon/Polygon';
import Rectangle from '../Icon/Rectangle';

export interface Props {
    /**
     * Unique ID for the component
     */
    id?: string;

    /**
     * Array of completed Tasks
     */
    tasks: TaskList['tasks'];

    /**
     * Allows to pass testid string for testing purposes
     */
    'data-testid'?: string;
}

const Container = styled.div`
    display: flex;
    flex-direction: column;
    align-items: center;
    gap: 0.5rem;
`;

const Title = styled.h1`
    text-align: center;
`;

const TasksList = styled.ul`
    display: grid;
    grid-template-columns: repeat(5, 1fr);
    gap: 0.5rem;
    list-style: none;
    margin: 0;
    padding: 0;
`;

const Task = styled.li`
    display: flex;
    flex-direction: column;
    justify-content: center;
    align-items: center;
    gap: 0.5rem;
    padding: 0.5rem;
    border-radius: 0.5rem;
    border: 1px dashed #335a6f;
`;

const CompletedTasks = ({ 'data-testid': dataTestId, ...props }: Props) => {
    // Similarly to Tasks component and depending on the use case, title value could be passed as prop
    const title = Boolean(props.tasks.length) ? 'Completed Tasks' : 'No Completed Tasks!';

    return (
        <Container id={props.id} data-testid={dataTestId}>
            <Title>{title}</Title>
            {Boolean(props.tasks.length) && (
                <TasksList>
                    {props.tasks.map(task => (
                        <Task key={task.id} id={task.id}>
                            {task.type === 'circle' && (
                                <Circle color="#82baa4" data-testid="icon-circle" />
                            )}
                            {task.type === 'rectangle' && (
                                <Rectangle color="#6491a9" data-testid="icon-rectangle" />
                            )}
                            {task.type === 'polygon' && (
                                <Polygon color="#bac1c4" data-testid="icon-polygon" />
                            )}
                            {task.name}
                        </Task>
                    ))}
                </TasksList>
            )}
        </Container>
    );
};

/** @component */
export default CompletedTasks;
