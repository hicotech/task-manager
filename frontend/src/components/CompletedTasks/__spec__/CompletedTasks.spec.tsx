import { render } from '@testing-library/react';
import type { Props as CompletedTasksProps } from '../CompletedTasks';
import CompletedTasks from '../CompletedTasks';

describe('Completed Tasks', () => {
    const testData: Required<CompletedTasksProps> & {
        name: string;
        noTasks: string;
    } = {
        id: 'completedTasks-id',
        name: 'Completed Tasks',
        noTasks: 'No Completed Tasks!',
        tasks: [
            {
                id: 'task-1',
                name: 'Task Circle',
                type: 'circle',
            },
            {
                id: 'task-2',
                name: 'Task Rectangle',
                type: 'rectangle',
            },
            {
                id: 'task-3',
                name: 'Task Polygon',
                type: 'polygon',
            },
        ],
        'data-testid': 'completedTasks-testid',
    };

    describe('Renders Default', () => {
        it('Render list name', () => {
            const { container, getByText, getByRole } = render(
                <CompletedTasks tasks={testData.tasks} />
            );

            expect(container).toMatchSnapshot();

            expect(getByRole('heading')).toBeInTheDocument();
            expect(getByText(testData.name)).toBeInTheDocument();
        });

        it('Render three tasks', () => {
            const { container, getByText } = render(<CompletedTasks tasks={testData.tasks} />);

            expect(container).toMatchSnapshot();

            expect(getByText(testData.tasks[0].name)).toBeInTheDocument();
            expect(getByText(testData.tasks[1].name)).toBeInTheDocument();
            expect(getByText(testData.tasks[2].name)).toBeInTheDocument();
        });

        it('Render two tasks', () => {
            const { container, getByText, queryByText } = render(
                <CompletedTasks tasks={testData.tasks.slice(0, 2)} />
            );

            expect(container).toMatchSnapshot();

            expect(getByText(testData.tasks[0].name)).toBeInTheDocument();
            expect(getByText(testData.tasks[1].name)).toBeInTheDocument();
            expect(queryByText(testData.tasks[2].name)).not.toBeInTheDocument();
        });

        it('Render one task', () => {
            const { container, getByText, queryByText } = render(
                <CompletedTasks tasks={testData.tasks.slice(0, 1)} />
            );

            expect(container).toMatchSnapshot();

            expect(getByText(testData.tasks[0].name)).toBeInTheDocument();
            expect(queryByText(testData.tasks[1].name)).not.toBeInTheDocument();
            expect(queryByText(testData.tasks[2].name)).not.toBeInTheDocument();
        });

        it('Render no tasks', () => {
            const { container, getByText, queryByText } = render(
                <CompletedTasks tasks={testData.tasks.slice(3)} />
            );

            expect(container).toMatchSnapshot();

            expect(getByText(testData.noTasks)).toBeInTheDocument();

            expect(queryByText(testData.name)).not.toBeInTheDocument();
            expect(queryByText(testData.tasks[0].name)).not.toBeInTheDocument();
            expect(queryByText(testData.tasks[1].name)).not.toBeInTheDocument();
            expect(queryByText(testData.tasks[2].name)).not.toBeInTheDocument();
        });
    });

    describe('Renders correct icons', () => {
        it('Render circle icon', () => {
            const { container, getByTestId } = render(
                <CompletedTasks tasks={testData.tasks.slice(0, 1)} />
            );

            expect(container).toMatchSnapshot();

            expect(getByTestId('icon-circle')).toBeInTheDocument();
        });

        it('Render rectangle icon', () => {
            const { container, getByTestId } = render(
                <CompletedTasks tasks={testData.tasks.slice(1, 2)} />
            );

            expect(container).toMatchSnapshot();

            expect(getByTestId('icon-rectangle')).toBeInTheDocument();
        });

        it('Render polygon icon', () => {
            const { container, getByTestId } = render(
                <CompletedTasks tasks={testData.tasks.slice(2, 3)} />
            );

            expect(container).toMatchSnapshot();

            expect(getByTestId('icon-polygon')).toBeInTheDocument();
        });
    });

    /**
     * Test tasks is skipped as it would duplicate tests from "Renders Default" section
     */
    describe('Props testing', () => {
        it('Id', () => {
            const { getByTestId } = render(
                <CompletedTasks
                    tasks={testData.tasks}
                    id={testData.id}
                    data-testid={testData['data-testid']}
                />
            );

            expect(getByTestId(testData['data-testid'])).toHaveAttribute('id', testData.id);
        });

        it('Data-testid', () => {
            const { getByTestId } = render(
                <CompletedTasks tasks={testData.tasks} data-testid={testData['data-testid']} />
            );

            expect(getByTestId(testData['data-testid'])).toBeInTheDocument();
        });
    });
});
