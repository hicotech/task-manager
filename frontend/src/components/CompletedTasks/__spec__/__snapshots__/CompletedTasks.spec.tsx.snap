// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Completed Tasks Renders Default Render list name 1`] = `
.c0 {
  display: flex;
  flex-direction: column;
  align-items: center;
  gap: 0.5rem;
}

.c1 {
  text-align: center;
}

.c2 {
  display: grid;
  grid-template-columns: repeat(5, 1fr);
  gap: 0.5rem;
  list-style: none;
  margin: 0;
  padding: 0;
}

.c3 {
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  gap: 0.5rem;
  padding: 0.5rem;
  border-radius: 0.5rem;
  border: 1px dashed #335a6f;
}

<div>
  <div
    class="c0"
  >
    <h1
      class="c1"
    >
      Completed Tasks
    </h1>
    <ul
      class="c2"
    >
      <li
        class="c3"
        id="task-1"
      >
        <svg
          data-testid="icon-circle"
          fill="#82baa4"
          height="1.5rem"
          viewBox="0 0 24 24"
          width="1.5rem"
          xmlns="http://www.w3.org/2000/svg"
        >
          <circle
            cx="12"
            cy="12"
            r="12"
          />
        </svg>
        Task Circle
      </li>
      <li
        class="c3"
        id="task-2"
      >
        <svg
          data-testid="icon-rectangle"
          fill="#6491a9"
          height="1.5rem"
          viewBox="0 0 24 24"
          width="1.5rem"
          xmlns="http://www.w3.org/2000/svg"
        >
          <rect
            height="24"
            width="24"
          />
        </svg>
        Task Rectangle
      </li>
      <li
        class="c3"
        id="task-3"
      >
        <svg
          data-testid="icon-polygon"
          fill="#bac1c4"
          height="1.5rem"
          viewBox="0 0 24 24"
          width="1.5rem"
          xmlns="http://www.w3.org/2000/svg"
        >
          <polygon
            points="12,0 24,8 19,24 4,24 0,8"
          />
        </svg>
        Task Polygon
      </li>
    </ul>
  </div>
</div>
`;

exports[`Completed Tasks Renders Default Render no tasks 1`] = `
.c0 {
  display: flex;
  flex-direction: column;
  align-items: center;
  gap: 0.5rem;
}

.c1 {
  text-align: center;
}

<div>
  <div
    class="c0"
  >
    <h1
      class="c1"
    >
      No Completed Tasks!
    </h1>
  </div>
</div>
`;

exports[`Completed Tasks Renders Default Render one task 1`] = `
.c0 {
  display: flex;
  flex-direction: column;
  align-items: center;
  gap: 0.5rem;
}

.c1 {
  text-align: center;
}

.c2 {
  display: grid;
  grid-template-columns: repeat(5, 1fr);
  gap: 0.5rem;
  list-style: none;
  margin: 0;
  padding: 0;
}

.c3 {
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  gap: 0.5rem;
  padding: 0.5rem;
  border-radius: 0.5rem;
  border: 1px dashed #335a6f;
}

<div>
  <div
    class="c0"
  >
    <h1
      class="c1"
    >
      Completed Tasks
    </h1>
    <ul
      class="c2"
    >
      <li
        class="c3"
        id="task-1"
      >
        <svg
          data-testid="icon-circle"
          fill="#82baa4"
          height="1.5rem"
          viewBox="0 0 24 24"
          width="1.5rem"
          xmlns="http://www.w3.org/2000/svg"
        >
          <circle
            cx="12"
            cy="12"
            r="12"
          />
        </svg>
        Task Circle
      </li>
    </ul>
  </div>
</div>
`;

exports[`Completed Tasks Renders Default Render three tasks 1`] = `
.c0 {
  display: flex;
  flex-direction: column;
  align-items: center;
  gap: 0.5rem;
}

.c1 {
  text-align: center;
}

.c2 {
  display: grid;
  grid-template-columns: repeat(5, 1fr);
  gap: 0.5rem;
  list-style: none;
  margin: 0;
  padding: 0;
}

.c3 {
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  gap: 0.5rem;
  padding: 0.5rem;
  border-radius: 0.5rem;
  border: 1px dashed #335a6f;
}

<div>
  <div
    class="c0"
  >
    <h1
      class="c1"
    >
      Completed Tasks
    </h1>
    <ul
      class="c2"
    >
      <li
        class="c3"
        id="task-1"
      >
        <svg
          data-testid="icon-circle"
          fill="#82baa4"
          height="1.5rem"
          viewBox="0 0 24 24"
          width="1.5rem"
          xmlns="http://www.w3.org/2000/svg"
        >
          <circle
            cx="12"
            cy="12"
            r="12"
          />
        </svg>
        Task Circle
      </li>
      <li
        class="c3"
        id="task-2"
      >
        <svg
          data-testid="icon-rectangle"
          fill="#6491a9"
          height="1.5rem"
          viewBox="0 0 24 24"
          width="1.5rem"
          xmlns="http://www.w3.org/2000/svg"
        >
          <rect
            height="24"
            width="24"
          />
        </svg>
        Task Rectangle
      </li>
      <li
        class="c3"
        id="task-3"
      >
        <svg
          data-testid="icon-polygon"
          fill="#bac1c4"
          height="1.5rem"
          viewBox="0 0 24 24"
          width="1.5rem"
          xmlns="http://www.w3.org/2000/svg"
        >
          <polygon
            points="12,0 24,8 19,24 4,24 0,8"
          />
        </svg>
        Task Polygon
      </li>
    </ul>
  </div>
</div>
`;

exports[`Completed Tasks Renders Default Render two tasks 1`] = `
.c0 {
  display: flex;
  flex-direction: column;
  align-items: center;
  gap: 0.5rem;
}

.c1 {
  text-align: center;
}

.c2 {
  display: grid;
  grid-template-columns: repeat(5, 1fr);
  gap: 0.5rem;
  list-style: none;
  margin: 0;
  padding: 0;
}

.c3 {
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  gap: 0.5rem;
  padding: 0.5rem;
  border-radius: 0.5rem;
  border: 1px dashed #335a6f;
}

<div>
  <div
    class="c0"
  >
    <h1
      class="c1"
    >
      Completed Tasks
    </h1>
    <ul
      class="c2"
    >
      <li
        class="c3"
        id="task-1"
      >
        <svg
          data-testid="icon-circle"
          fill="#82baa4"
          height="1.5rem"
          viewBox="0 0 24 24"
          width="1.5rem"
          xmlns="http://www.w3.org/2000/svg"
        >
          <circle
            cx="12"
            cy="12"
            r="12"
          />
        </svg>
        Task Circle
      </li>
      <li
        class="c3"
        id="task-2"
      >
        <svg
          data-testid="icon-rectangle"
          fill="#6491a9"
          height="1.5rem"
          viewBox="0 0 24 24"
          width="1.5rem"
          xmlns="http://www.w3.org/2000/svg"
        >
          <rect
            height="24"
            width="24"
          />
        </svg>
        Task Rectangle
      </li>
    </ul>
  </div>
</div>
`;

exports[`Completed Tasks Renders correct icons Render circle icon 1`] = `
.c0 {
  display: flex;
  flex-direction: column;
  align-items: center;
  gap: 0.5rem;
}

.c1 {
  text-align: center;
}

.c2 {
  display: grid;
  grid-template-columns: repeat(5, 1fr);
  gap: 0.5rem;
  list-style: none;
  margin: 0;
  padding: 0;
}

.c3 {
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  gap: 0.5rem;
  padding: 0.5rem;
  border-radius: 0.5rem;
  border: 1px dashed #335a6f;
}

<div>
  <div
    class="c0"
  >
    <h1
      class="c1"
    >
      Completed Tasks
    </h1>
    <ul
      class="c2"
    >
      <li
        class="c3"
        id="task-1"
      >
        <svg
          data-testid="icon-circle"
          fill="#82baa4"
          height="1.5rem"
          viewBox="0 0 24 24"
          width="1.5rem"
          xmlns="http://www.w3.org/2000/svg"
        >
          <circle
            cx="12"
            cy="12"
            r="12"
          />
        </svg>
        Task Circle
      </li>
    </ul>
  </div>
</div>
`;

exports[`Completed Tasks Renders correct icons Render polygon icon 1`] = `
.c0 {
  display: flex;
  flex-direction: column;
  align-items: center;
  gap: 0.5rem;
}

.c1 {
  text-align: center;
}

.c2 {
  display: grid;
  grid-template-columns: repeat(5, 1fr);
  gap: 0.5rem;
  list-style: none;
  margin: 0;
  padding: 0;
}

.c3 {
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  gap: 0.5rem;
  padding: 0.5rem;
  border-radius: 0.5rem;
  border: 1px dashed #335a6f;
}

<div>
  <div
    class="c0"
  >
    <h1
      class="c1"
    >
      Completed Tasks
    </h1>
    <ul
      class="c2"
    >
      <li
        class="c3"
        id="task-3"
      >
        <svg
          data-testid="icon-polygon"
          fill="#bac1c4"
          height="1.5rem"
          viewBox="0 0 24 24"
          width="1.5rem"
          xmlns="http://www.w3.org/2000/svg"
        >
          <polygon
            points="12,0 24,8 19,24 4,24 0,8"
          />
        </svg>
        Task Polygon
      </li>
    </ul>
  </div>
</div>
`;

exports[`Completed Tasks Renders correct icons Render rectangle icon 1`] = `
.c0 {
  display: flex;
  flex-direction: column;
  align-items: center;
  gap: 0.5rem;
}

.c1 {
  text-align: center;
}

.c2 {
  display: grid;
  grid-template-columns: repeat(5, 1fr);
  gap: 0.5rem;
  list-style: none;
  margin: 0;
  padding: 0;
}

.c3 {
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  gap: 0.5rem;
  padding: 0.5rem;
  border-radius: 0.5rem;
  border: 1px dashed #335a6f;
}

<div>
  <div
    class="c0"
  >
    <h1
      class="c1"
    >
      Completed Tasks
    </h1>
    <ul
      class="c2"
    >
      <li
        class="c3"
        id="task-2"
      >
        <svg
          data-testid="icon-rectangle"
          fill="#6491a9"
          height="1.5rem"
          viewBox="0 0 24 24"
          width="1.5rem"
          xmlns="http://www.w3.org/2000/svg"
        >
          <rect
            height="24"
            width="24"
          />
        </svg>
        Task Rectangle
      </li>
    </ul>
  </div>
</div>
`;
